lib/SIL/SILVerifier.cpp

```diff
@@ -1254,6 +1254,36 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
   }
 
   void checkAutoDiffFunctionInst(AutoDiffFunctionInst *adfi) {
+    // If the associated function comes from a reabstraction thunk, then it is
+    // impossible to determine the type of the associated function from the
+    // type of the original function, because we also need to know the
+    // abstraction pattern that the reabstraction made. So we currently
```

Comment:
I see what's going on here:

The autodiff-associated function type calculation produces the following type:
```swift
$@noescape @callee_guaranteed (@in_guaranteed Float) -> (@out Float, @owned @callee_guaranteed (Float) -> Float)
```
But reabstraction makes the closure result's parameters and results indirect:
```swift
$@noescape @callee_guaranteed (@in_guaranteed Float) -> (@out Float, @owned @callee_guaranteed (@in_guaranteed Float) -> @out Float)
```

Currently `SILFunctionType::getAutoDiffAssociatedFunctionType` uses `getFormalResultInfo` and `getFormalParameterInfo` to determine the parameter convention and result convention of the differential/pullback.  If we instead mirror their conventions in the original function type, then associated fn types can still be correctly calculated even if they are reabstracted, I think.

Reply:
I updated `SILFunctionType::getAutoDiffAssociatedFunctionType` in https://github.com/apple/swift/commit/372748eef91bf3335fa1a5f6655d73670cb48780 and it's passing all tests. Feel free to pull that in.